### PR TITLE
Finalize nation select procedure and fix animated images

### PR DESCRIPTION
**Pull request by stevewgr**
### Description

nation-select

Ignore the mouse cursor there is a bit off track because of OBS recording software.

This PR finalizes the nation select procedure with some additional fixes:
+ Remove confirmation when selecting nation.
+ Fix fitting description text in nation select proc.
+ Remove triggering audio from login to nation proc.
+ Fix funky bug with animated images.
+ Fix audio in login proc.
+ Fix going back from nation to login procedure.
+ Trigger audio for create character procedure.
+ Tweak audio transition in nation select proc.
+ Define audio constants instead of magic values.
+ Improve nation change transition in nation select.
+ Use CN3SndObjStream instead of CN3SndObj.
+ Reorder CUINationSelectDlg class and apply renames.

### CODE REVIEW

**Review by GiRx8 — APPROVED** [review]
It's working well! just the transition between nation swap seems a bit slow (don't know if it was always like that back that)

File: src/engine/N3Base/N3UIImage.cpp [hunk 1]
```diff
@@ -35,6 +35,7 @@ CN3UIImage::~CN3UIImage() {
         delete[] m_pAnimImagesRef;
         m_pAnimImagesRef = NULL;
     }
+    m_szTexFN = "";
```
Comment (ko4life-psko) [inline src/engine/N3Base/N3UIImage.cpp:38]: I think you can remove this from the destructor. In the Release function it makes sense in case one wants to release existing resources, but not destroy the object. However, in a destructor since `std::string` does not need manual deallocation, then it's not needed.
```suggestion
```

File: src/engine/N3Base/N3UIImage.h [hunk 2]
```diff
@@ -29,7 +29,7 @@ class CN3UIImage : public CN3UIBase {
 
     float         m_fAnimFrame;     // 1�ʴ� �׷���� �̹����� ��
     int           m_iAnimCount;     // Animate�Ǵ� �̹��� ��(Load�� �� child�� ���ڷ� �ľ�)
-    int           m_fCurAnimFrame;  // ���� �׷��� animate frame index (�Ҽ�� ��� �������)
+    float         m_fCurAnimFrame;  // ���� �׷��� animate frame index (�Ҽ�� ��� �������)
```
Comment (ko4life-psko) [inline src/engine/N3Base/N3UIImage.h:32]: That's need a funky one. I believe they were testing with it officially, the fact that the variable was starting with `f` to indicate a `float`, but was declared as `int`. Good catch and important fix it is!

File: src/game/UINationSelectDlg.cpp [hunk 3]
```diff
@@ -333,3 +216,104 @@ void CUINationSelectDlg::Render() {
...
+bool CUINationSelectDlg::ReceiveMessage(CN3UIBase * pSender, DWORD dwMsg) {
+    if (!pSender) {
+        return false;
+    }
+
+    if (!m_bTransitionActive && dwMsg == UIMSG_BUTTON_CLICK) {
+        if (pSender == (CN3UIBase *)m_pBtnCloseKa || pSender == (CN3UIBase *)m_pBtnCloseEl) {
+            CGameProcedure::s_pSocket->Disconnect();
+            CGameProcedure::ProcActiveSet((CGameProcedure *)CGameProcedure::s_pProcLogIn);
+            CGameProcedure::s_pProcLogIn->m_bLogIn = true;
```
Comment (ko4life-psko) [inline src/game/UINationSelectDlg.cpp:229]: Curious question; why do we set here `m_bLogIn` to `true` when exiting the nation select procedure? Especially after exiting means that we're logged out. I suppose to indicate that the user is passed through the authentication stage if they are at the nation select procedure? In that case, maybe we should set it from the login procedure instead when we receive a successful authentication message from the server.

File: src/game/UINationSelectDlg.cpp [hunk 4]
```diff
@@ -333,3 +216,104 @@ void CUINationSelectDlg::Render() {
...
+        ::_LoadStringFromResource(IDS_NATIONSELECT_DESCRIPTION_EL, szBuff);
+        m_pTextNoticeEl->SetString(szBuff);
+        m_pTextNoticeEl->SetFont(m_pTextNoticeEl->GetFontName(), m_pTextNoticeEl->GetFontHeight() - 2, FALSE, FALSE);
+    }
+    if (m_pTextNoticeKa) {
+        // Karus is a new nation that was established in Adonis...
+        ::_LoadStringFromResource(IDS_NATIONSELECT_DESCRIPTION_KA, szBuff);
+        m_pTextNoticeKa->SetString(szBuff);
+        m_pTextNoticeKa->SetFont(m_pTextNoticeKa->GetFontName(), m_pTextNoticeKa->GetFontHeight() - 2, FALSE, FALSE);
+    }
```
Comment (ko4life-psko) [inline src/game/UINationSelectDlg.cpp:258]: Is it how it behaves officially with getting the font size and - 2? I haven't checked, but I believe if the font doesn't fit in the description box, maybe it's using the wrong font or text wrapping isn't working correctly.

In a situation we still need to decrease the font size, as much as it's not ideal, I suggest modifying the relevant game asset UIF file to reduce confusion when users set font size in the UIF files, but it doesn't reflect correctly in game.
```suggestion
    if (m_pTextNoticeEl) {
        // El Morad is a nation that was formed by King Manes...
        ::_LoadStringFromResource(IDS_NATIONSELECT_DESCRIPTION_EL, szBuff);
        m_pTextNoticeEl->SetString(szBuff);
    }
    if (m_pTextNoticeKa) {
        // Karus is a new nation that was established in Adonis...
        ::_LoadStringFromResource(IDS_NATIONSELECT_DESCRIPTION_KA, szBuff);
        m_pTextNoticeKa->SetString(szBuff);
```

File: src/game/UINationSelectDlg.cpp [hunk 5]
```diff
@@ -333,3 +216,104 @@ void CUINationSelectDlg::Render() {
...
+void CUINationSelectDlg::ChangeNation(bool bNationKa) {
+    if (bNationKa) {
+        m_bNationKa = true;
+        if (m_pSndEl) {
+            m_pSndEl->Stop(3.0f);
+        }
+        if (m_pSndKa) {
+            m_pSndKa->Looping(true);
+            m_pSndKa->Play(NULL, 0.0f, 6.0f);
+        }
```
Comment (ko4life-psko) [inline src/game/UINationSelectDlg.cpp:292]: From what I've seen, we always set this audio to looping `true`. I suggest moving this into `CUINationSelectDlg::InitResources` for both KA and EL.
```suggestion
        if (m_pSndKa) {
            m_pSndKa->Play(NULL, 0.0f, 6.0f);
        }
```

**Review by ko4life-psko — APPROVED** [review]
Awesome work, @stevewgr! This nation select screen definitely brings some good memories for me, thank you for that!

I tested it locally and everything seems to be working as expected. I left some feedback for you to review.

From my end this PR is approved, since my feedback is more about nit-picking rather than required changes.

**Comment by ko4life-psko** [comment]
> It's working well! just the transition between nation swap seems a bit slow (don't know if it was always like that back that)

@GiRx8, I personally haven't checked, but if I remember correctly, officially it was even slower. Hard to remember for sure. @stevewgr, is that true?

I guess if you throw here the v1264 client exe and I might have time to have a quick peek.

File: src/game/UINationSelectDlg.cpp (hunk 4 above)
Comment (stevewgr) [reply src/game/UINationSelectDlg.cpp:258]: I checked the assembly and saw that they don't do that. I agree with your thoughts. Here is a PR that fixes the actual file: https://github.com/ko4life-net/ko-assets/pull/6

File: src/engine/N3Base/N3UIImage.h (hunk 2 above)
Comment (stevewgr) [reply src/engine/N3Base/N3UIImage.h:32]: Thank you! It was an annoying issue to debug indeed.

File: src/game/UINationSelectDlg.cpp (hunk 3 above)
Comment (stevewgr) [reply src/game/UINationSelectDlg.cpp:229]: Good question. It's how it's done officially in the v1264 client:
```c
  if ( !this->m_bTransitionActive && a_dwMsg == 1 )
  {
    if ( a_pSender == (CN3UIBase *)this->m_pBtnCloseKa )
    {
LABEL_6:
      CAPISocket::Disconnect(s_pSocket);
      CGameProcedure::ProcActiveSet((int)s_pProcLogIn);
      BYTE1(s_pProcLogIn->m_bLogIn) = 1;
      return 1;
    }
```

I was looking for references in the existing code and `m_bLogIn` is nowhere to be used on the login procedure level apart from assigning true or false. So I spend some time checking the assembly and the only place it was used is in `CUILogIn::ReceiveMessage`:
```c
....
  if ( a_dwMsg != 0x1000 )                      // UIMSG_EDIT_RETURN
  {
    if ( a_dwMsg == 0x200000 && a_pSender == (CN3UIBase *)this->m_ListServerInfos.heap )// UIMSG_LIST_SELCHANGE
      CUILogIn::ServerInfoUpdate(this);
    return 1;
  }
  v12 = s_pProcLogIn;
  if ( BYTE1(s_pProcLogIn->m_bLogIn) || LOBYTE(this->m_bLogIn) || !this->m_pEdit_id || !this->m_pEdit_pw )
  {
    if ( !LOBYTE(this->m_bOpenningNow) )
    {
      m_pGroup_ServerList = this->m_pGroup_ServerList;
      if ( m_pGroup_ServerList )
      {
        if ( LOBYTE(m_pGroup_ServerList->CN3UIBase.m_bVisible) )
        {
          pMsgBox = s_pMsgBoxMgr->m_pMsgBoxLatestRef;
          if ( !pMsgBox || !LOBYTE(pMsgBox->CN3UIBase.m_bVisible) )
          {
            CGameProcLogIn::ConnectToGameServer(s_pProcLogIn);
            v12 = s_pProcLogIn;
          }
          BYTE1(v12->m_bLogIn) = 0;
        }
      }
    }
    return 1;
  }
  m_pMsgBoxLatestRef = s_pMsgBoxMgr->m_pMsgBoxLatestRef;
  if ( !m_pMsgBoxLatestRef || !LOBYTE(m_pMsgBoxLatestRef->CN3UIBase.m_bVisible) )
    goto LABEL_7;
  return 1;
}
```

I suppose it's used to ensure globally (considering the static variable of the procedure is assigned) to not send login packets multiple times.
However I was not able to reproduce, since there is already existing check in code:
```c
    } else if (dwMsg == UIMSG_EDIT_RETURN) {
        if (!m_bLogIn && m_pEdit_id && m_pEdit_pw) {
            CN3UIBase * pMsgBox = CGameProcedure::s_pMsgBoxMgr->GetFocusMsgBox();
            if (!(pMsgBox && pMsgBox->IsVisible())) {
                CGameProcedure::s_pProcLogIn->MsgSend_AccountLogIn(LIC_KNIGHTONLINE);
            }
        }
    }
```

Long story short. I assign it there just to keep it as in the official client, but still unsure what scenario it covers.

File: src/game/UINationSelectDlg.cpp (hunk 4 above)
Comment (stevewgr) [reply src/game/UINationSelectDlg.cpp]: Also see related commit: https://github.com/ko4life-net/ko/pull/125/commits/cd13eb401fcfefd301d4809ff7b1944999c91861

File: src/engine/N3Base/N3UIImage.cpp (hunk 1 above)
Comment (stevewgr) [reply src/engine/N3Base/N3UIImage.cpp]: Thanks for the info! Agree. Please see commit: https://github.com/ko4life-net/ko/pull/125/commits/e9f6f91a35b77135f861e57c5d5affc5929c8ba4

File: src/game/UINationSelectDlg.cpp (hunk 5 above)
Comment (stevewgr) [reply src/game/UINationSelectDlg.cpp]: Good point! Please see commit: https://github.com/ko4life-net/ko/pull/125/commits/2fc84fbf2d675725b24d26cc7242fda4218120c2

**Comment by stevewgr** [comment]
Hey guys, thanks for the great feedback!

> > It's working well! just the transition between nation swap seems a bit slow (don't know if it was always like that back that)
> 
> @GiRx8, I personally haven't checked, but if I remember correctly, officially it was even slower. Hard to remember for sure. @stevewgr, is that true?

Yes, officially it was much slower by design I suppose. The actual transition speed is implemented in the v1264 client at address `0x005BCE90`, which is `CUINationSelectDlg::Render`:
```c++
...
    m_fTextureFactor -= s_fSecPerFrm * 0.3333333f;
    if (m_fTextureFactor < 0.30f) {
        m_fTextureFactor = 1.0f;
        m_bTransitionActive = false;
....
```
Officially the value `0.30f` is `0.0f`. I set it to `0.30f` to make it a bit faster. Making it faster than this I think is border line custom behavior that ideally we avoid in order to stick to the official behavior.

> I guess if you throw here the v1264 client exe and I might have time to have a quick peek.

Just in case anyone needs unpacked v1264 client: 
[KnightOnLine_1264.7z](https://github.com/ko4life-net/ko/files/9742219/KnightOnLine_1264.txt)

GitHub doesn't allow to upload `*.7z` extension file, so I renamed it to `*.txt` just for uploading and password is: `stevewgr`.

File: src/engine/N3Base/N3UIImage.cpp (hunk 1 above)
Comment (ko4life-psko) [reply src/engine/N3Base/N3UIImage.cpp]: Glad to see someone actually using git [--fixup](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---fixupamendrewordltcommitgt) commits. This makes the review process so pleasant. LGTM.

File: src/game/UINationSelectDlg.cpp (hunk 3 above)
Comment (ko4life-psko) [reply src/game/UINationSelectDlg.cpp:229]: Makes sense. Many thanks for the detailed explanation and for spending the time checking this.

File: src/game/UINationSelectDlg.cpp (hunk 4 above)
Comment (ko4life-psko) [reply src/game/UINationSelectDlg.cpp]: Good change. Nice work! 🚀

**Review by ko4life-psko — APPROVED** [review]
Many thanks for the changes, @stevewgr! LGTM. 🚀
